src/Polysemy/Final.hs

```diff
@@ -208,77 +169,86 @@ liftS m = do
 -- | Embed a kleisli action into the stateful environment, in terms of the final
 -- monad. You can use 'bindS' to get an effect parameter of the form @a -> n b@
 -- into something that can be used after calling 'runS' on an effect parameter
 -- @n a@.
-bindS :: Monad m => (a -> n b) -> Sem (WithStrategy m f n) (f a -> m (f b))
-bindS = fmap (runM .) . bindT
+bindS :: (a -> n b) -> Sem (WithStrategy m f n) (f a -> m (f b))
+bindS = send . HoistInterpretation
 {-# INLINE bindS #-}
 
 ------------------------------------------------------------------------------
--- | Internal function to process Strategies in terms of 'withWeaving'.
-runStrategy :: Functor f
-            => f ()
-            -> (forall x. f (n x) -> m (f x))
-            -> (forall x. f x -> Maybe x)
-            -> Sem (WithStrategy m f n) a
-            -> a
-runStrategy s wv ins (Sem m) = runIdentity $ m $ \u -> case extract u of
-  Weaving e s' _ ex' _ -> Identity $ ex' $ (<$ s') $ case e of
-    GetInitialState -> s
-    HoistInterpretation na -> embed . wv . fmap na
-    GetInspector -> Inspector ins
-{-# INLINE runStrategy #-}
-
-------------------------------------------------------------------------------
--- | Lower a 'Sem' containing only a lifted, final monad into that monad.
--- The appearance of 'Lift' as the final effect
--- is to allow the use of operations that rely on a @'LastMember' ('Lift' m)@
--- constraint.
-runFinal :: Monad m => Sem '[Final m, Embed m] a -> m a
-runFinal = usingSem $ \u -> case decomp u of
-  Right (Weaving (WithWeaving wav) s wv ex ins) ->
+-- | Lower a 'Sem' containing only a single lifted, final 'Monad' into that
+-- monad.
+runFinal :: Monad m => Sem '[Final m] a -> m a
+runFinal = usingSem $ \u -> case extract u of
+  Weaving (WithWeaving wav) s wv ex ins ->
     ex <$> wav s (runFinal . wv) ins
-  Left g -> case extract g of
-    Weaving (Embed m) s _ ex _ -> ex . (<$ s) <$> m
 {-# INLINE runFinal #-}
 
 ------------------------------------------------------------------------------
--- | Lower a 'Sem' containing two lifted monad into the final monad,
--- by interpreting one of the monads in terms of the other one.
---
--- This allows for the use of operations that rely on a @'LastMember' ('Lift' m)@
--- constraint, as long as @m@ can be transformed to the final monad;
--- but be warned, this breaks the implicit contract of @'LastMember' ('Lift' m)@
--- that @m@ /is/ the final monad, so depending on the final monad and operations
--- used, 'runFinalLift' may become /unsafe/.
+-- | Run a 'Sem' containing only @'Final' 'Identity'@ as a pure value.
+runFinalPure :: forall a. Sem '[Final Identity] a -> a
+runFinalPure = coerce (runFinal @Identity @a)
+{-# INLINE runFinalPure #-}
+
+------------------------------------------------------------------------------
+-- | Given natural transformations between @m1@ and @m2@, run a @'Final' m1@
+-- effect by transforming it into a @'Final' m2@ effect.
+finalToFinal :: forall m1 m2 a r
```

Comment:
What's the use case for this? The isomorphism between `m1` and `m2` restricts any case I can think of.

Reply:
What I require here isn't an isomorphism, but I digress: an example would be between `m` and `Codensity m`. `lowerFinal` is also interpreted in terms of `finalToFinal` (but it doesn't need to be).

Reply:
Fair enough, thanks!